apollo-router/src/query_planner/rewrites.rs

```diff
@@ -0,0 +1,94 @@
+//! to be future-proof by supporting all types of rewrites on both "sides".
+
+use serde::Deserialize;
+use serde::Serialize;
+
+use crate::json_ext::Path;
+use crate::json_ext::PathElement;
+use crate::json_ext::Value;
+use crate::json_ext::ValueExt;
+use crate::spec::Schema;
+
+/// Given a path, separates the last element of path and the rest of it and return them as a pair.
+/// This will return `None` if the path is empty.
+fn split_path_last_element(path: &Path) -> Option<(Path, &PathElement)> {
+    // If we have a `last()`, then we have a `parent()` too, so unwrapping shoud be safe.
+    path.last().map(|last| (path.parent().unwrap(), last))
+}
+
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
+#[serde(rename_all = "PascalCase", tag = "kind")]
+pub(crate) enum DataRewrite {
+    ValueSetter(DataValueSetter),
+    KeyRenamer(DataKeyRenamer),
+}
+
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct DataValueSetter {
+    pub(crate) path: Path,
+    pub(crate) set_value_to: Value,
+}
+
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct DataKeyRenamer {
+    pub(crate) path: Path,
+    pub(crate) rename_key_to: String,
+}
+
+impl DataRewrite {
+    fn maybe_apply(&self, schema: &Schema, data: &mut Value) {
+        match self {
+            DataRewrite::ValueSetter(setter) => {
+                // The `path` of rewrites can only be either `Key` or `Fragment`, and so far
+                // we only ever rewrite the value of fields, so the last element will be a
+                // `Key` and we ignore other cases (in theory, it could be `Fragment` needs
+                // to be supported someday if we ever need to rewrite full object values,
+                // but that can be added then).
+                if let Some((parent, PathElement::Key(k))) = split_path_last_element(&setter.path) {
+                    data.select_values_and_paths_mut(schema, &parent, |_path, obj| {
+                        if let Some(value) = obj.get_mut(k) {
+                            *value = setter.set_value_to.clone()
+                        }
+                    });
+                }
+            }
+            DataRewrite::KeyRenamer(renamer) => {
+                // As the name implies, this only applies to renaming "keys", so we're
+                // guaranteed the last element is one and can ignore other cases.
+                if let Some((parent, PathElement::Key(k))) = split_path_last_element(&renamer.path)
+                {
+                    data.select_values_and_paths_mut(schema, &parent, |_path, selected| {
+                        if let Some(obj) = selected.as_object_mut() {
+                            if let Some(value) = obj.remove(k.as_str()) {
+                                obj.insert(renamer.rename_key_to.clone(), value);
+                            }
```

Comment:
```suggestion
                            if obj.contains_key(k.as_str()) {
                                obj.insert(renamer.rename_key_to.clone(), value);
                            }
```

a remove then insert would change the order of keys in the object (which should be fixed by response formatting later anyway, but reordering elements in an indexmap can be quite expensive)

Reply:
I agree that remove-then-insert does change the order and in an ideal world it would nice not too.

But I'm not sure about your suggestion: it leaves the old entry that is meant to be "renamed" in the response, and I actually don't think that's safe to do (and even if it was now, I'd argue that it's a bit dodgy as future renames could also rely on the renamed value being genuinely removed).

Ideally, we'd be able to reuse the index of the removed entry for the replacing entry (thus keeping the ordering and saving for reordering), but I don't see how to do that on the `serde_json_bytes::Map` we have here (it seems that would be doable if we directly had an [IndexMap](https://docs.rs/indexmap/latest/indexmap/map/struct.IndexMap.html) through some combination of get stuffs by index, inserting and then using `swap_remove_index`, but none of this seems exposed here unless I missed something). It's possible I'm missing some way to do this.

That said, I'm also not sure there is a performance concern here: `obj.remove` seems to be calling `swap_remove` on the underlying index map, so while the order of element is perturbed, there is no expensive re-assignment of indexes (it's O(1) on average according to the doc). And as you mentioned, perturbing the order has no real consequences since everything gets re-inserted in order during response formatting.

Reply:
alright, let's leave it like that then